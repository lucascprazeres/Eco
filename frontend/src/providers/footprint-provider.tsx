import { useMutation } from '@apollo/client'
import { CALCULATE_CARBON_FOOTPRINT } from '@eco/graphql/mutation'
import {
  CalculateFootprintMutationOutput,
  CarbonFootprint,
  CarbonFootprintInput,
} from '@eco/models/carbon-footprint'
import {
  createContext,
  PropsWithChildren,
  useCallback,
  useContext,
  useState,
} from 'react'

interface FootprintContextData {
  footprint: CarbonFootprint | null
  handleCalculateFootprint: (data: CarbonFootprintInput) => Promise<void>
}

const FootprintContext = createContext<FootprintContextData>(
  {} as FootprintContextData,
)

export function FootprintProvider({ children }: PropsWithChildren) {
  const [footprint, setFootprint] = useState<CarbonFootprint | null>(null)

  const [calculateFootprint] = useMutation<CalculateFootprintMutationOutput>(
    CALCULATE_CARBON_FOOTPRINT,
  )

  const handleCalculateFootprint = useCallback(
    async (formData: CarbonFootprintInput) => {
      const { data } = await calculateFootprint({
        variables: {
          input: formData,
        },
      })

      setFootprint(data?.calculateFootprint as unknown as CarbonFootprint)
    },
    [calculateFootprint],
  )

  return (
    <FootprintContext.Provider value={{ footprint, handleCalculateFootprint }}>
      {children}
    </FootprintContext.Provider>
  )
}

export const useFootprint = () => useContext(FootprintContext)
