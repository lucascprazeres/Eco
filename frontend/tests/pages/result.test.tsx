import { render, screen, fireEvent, cleanup } from '@testing-library/react'
import { describe, it, vi, expect, beforeEach } from 'vitest'
import Result from '@eco/pages/result'
import { useFootprint } from '@eco/providers/footprint-provider'
import { footprintMock } from '../mocks/footprint'
import mockRouter from 'next-router-mock'

vi.mock('@eco/providers/footprint-provider', () => ({
  useFootprint: vi.fn(),
}))

vi.mock('next/router', () => vi.importActual('next-router-mock'))

vi.mock('@eco/components/footprint-table', () => ({
  FootprintTable: () => <div>Mocked FootprintTable</div>,
}))
vi.mock('@eco/components/footprint-pie-chart', () => ({
  FootprintPieChart: () => <div>Mocked FootprintPieChart</div>,
}))

describe('Result Page', () => {
  beforeEach(() => {
    cleanup()
    vi.clearAllMocks()
  })

  it('should be able to redirect user if footprint is empty', () => {
    // arrange
    // @ts-expect-error disabled warning because typescript doesn't recognize the hook is mocked
    useFootprint.mockReturnValue({
      footprint: null,
    })

    // act
    render(<Result />)

    // assert
    expect(mockRouter).toMatchObject({
      pathname: '/',
    })
  })

  it('should render the page with a footprint', () => {
    // arrange
    // @ts-expect-error disabled warning because typescript doesn't recognize the hook is mocked
    useFootprint.mockReturnValue({
      footprint: footprintMock,
    })

    // act
    const result = render(<Result />)

    // assert
    expect(
      screen.getByText('This is your Carbon Footprint'),
    ).toBeInTheDocument()

    expect(screen.getByText('Mocked FootprintTable')).toBeInTheDocument()

    expect(screen.getByText('Mocked FootprintPieChart')).toBeInTheDocument()
    expect(
      screen.getByText('Congratulations! You took the first step'),
    ).toBeInTheDocument()

    expect(result).toMatchSnapshot()
  })

  it('should navigate to CALCULATOR_PAGE when "Try again!" button is clicked', () => {
    // arrange
    // @ts-expect-error disabled warning because typescript doesn't recognize the hook is mocked
    useFootprint.mockReturnValue({
      footprint: footprintMock,
    })

    // act
    render(<Result />)

    fireEvent.click(screen.getByText('Try again!'))

    // assert
    expect(mockRouter).toMatchObject({
      pathname: '/calculator',
    })
  })
})
