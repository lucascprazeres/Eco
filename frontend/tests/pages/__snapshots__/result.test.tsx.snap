// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Result Page > should render the page with a footprint 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-21eb5z"
      >
        <header
          class="MuiBox-root css-1utcncn"
        >
          <div
            class="MuiBox-root css-axw7ok"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-zg625t-MuiTypography-root"
            >
              Eco
            </h4>
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="40"
              loading="lazy"
              src="/src/assets/logo.svg"
              style="color: transparent;"
              width="40"
            />
          </div>
        </header>
        <main
          class="MuiBox-root css-16too2"
        >
          <div
            class="MuiBox-root css-1esezgf"
          >
            <div
              class="MuiBox-root css-axw7ok"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 css-ve8gwu-MuiTypography-root"
              >
                This is your Carbon Footprint
              </h5>
              <img
                alt=""
                data-nimg="1"
                decoding="async"
                height="32"
                loading="lazy"
                src="/src/assets/smile.svg"
                style="color: transparent;"
                width="32"
              />
            </div>
            <div
              class="MuiBox-root css-1lekzkb"
            >
              <div>
                Mocked FootprintPieChart
              </div>
              <div>
                Mocked FootprintTable
              </div>
            </div>
            <div
              class="MuiBox-root css-0"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-1ghywoa-MuiTypography-root"
              >
                Congratulations! You took the first step
              </h6>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1m3a417-MuiTypography-root"
              >
                Your carbon footprint is a snapshot of your current environmental impact. Small changes in your daily habits—like using public transport, reducing energy consumption, or cutting food waste—can make a big difference. Try adjusting your behavior and come back to see how your footprint improves.
              </p>
            </div>
            <div
              class="MuiBox-root css-17ea04y"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-fq2dir-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Try again!
              </button>
            </div>
          </div>
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-21eb5z"
    >
      <header
        class="MuiBox-root css-1utcncn"
      >
        <div
          class="MuiBox-root css-axw7ok"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-zg625t-MuiTypography-root"
          >
            Eco
          </h4>
          <img
            alt=""
            data-nimg="1"
            decoding="async"
            height="40"
            loading="lazy"
            src="/src/assets/logo.svg"
            style="color: transparent;"
            width="40"
          />
        </div>
      </header>
      <main
        class="MuiBox-root css-16too2"
      >
        <div
          class="MuiBox-root css-1esezgf"
        >
          <div
            class="MuiBox-root css-axw7ok"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 css-ve8gwu-MuiTypography-root"
            >
              This is your Carbon Footprint
            </h5>
            <img
              alt=""
              data-nimg="1"
              decoding="async"
              height="32"
              loading="lazy"
              src="/src/assets/smile.svg"
              style="color: transparent;"
              width="32"
            />
          </div>
          <div
            class="MuiBox-root css-1lekzkb"
          >
            <div>
              Mocked FootprintPieChart
            </div>
            <div>
              Mocked FootprintTable
            </div>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-1ghywoa-MuiTypography-root"
            >
              Congratulations! You took the first step
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body1 css-1m3a417-MuiTypography-root"
            >
              Your carbon footprint is a snapshot of your current environmental impact. Small changes in your daily habits—like using public transport, reducing energy consumption, or cutting food waste—can make a big difference. Try adjusting your behavior and come back to see how your footprint improves.
            </p>
          </div>
          <div
            class="MuiBox-root css-17ea04y"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-fq2dir-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Try again!
            </button>
          </div>
        </div>
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
